File: instrumentation/common-api/src/main/java/io/opentelemetry/android/instrumentation/common/ScreenNameExtractor.java
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.android.instrumentation.common;

import io.opentelemetry.android.instrumentation.annotations.RumScreenName;

public interface ScreenNameExtractor {

    String extract(Object instance);

    ScreenNameExtractor DEFAULT =
            new ScreenNameExtractor() {
                @Override
                public String extract(Object instance) {
                    return useAnnotationOrClassName(instance.getClass());
                }

                private String useAnnotationOrClassName(Class<?> clazz) {
                    RumScreenName rumScreenName = clazz.getAnnotation(RumScreenName.class);
                    return rumScreenName == null ? clazz.getSimpleName() : rumScreenName.value();
                }
            };
}

File: instrumentation/common-api/src/main/java/io/opentelemetry/android/instrumentation/common/ScreenNameExtractor.kt
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.android.instrumentation.common

fun interface ScreenNameExtractor {
    /**
     * Obtains a screen name from the supplied instance.
     */
    fun extract(instance: Any): String
}
